pkg/issuer/acme/dns/azuredns/azuredns.go

```diff
@@ -84,19 +111,26 @@ func getAuthorization(env azure.Environment, clientID, clientSecret, subscriptio
 		}
 		return spt, nil
 	}
-	logf.Log.V(logf.InfoLevel).Info("No ClientID found:  authenticating azuredns with managed identity (MSI)")
+	logf.Log.V(logf.InfoLevel).Info("No ClientID found: attempting to authenticate with ambient credentials (Azure Workload Identity or Azure Managed Identity, in that order)")
+
 	if !ambient {
-		return nil, fmt.Errorf("ClientID is not set but neither `--cluster-issuer-ambient-credentials` nor `--issuer-ambient-credentials` are set. These are necessary to enable Azure Managed Identities")
+		return nil, fmt.Errorf("ClientID is not set but neither `--cluster-issuer-ambient-credentials` nor `--issuer-ambient-credentials` are set. These are necessary to enable ambient credentials")
+	}
+
+	spt, err := getWorkloadIdentity(env)
+	if spt != nil {
+		return spt, nil
 	}
```

Comment:
This `err` should be checked and returned.

Reply:
Just to add to that, there could be multiple reasons why a token cannot be obtained. E.g. I could easily imagine a case where someone would forget to create a federated-credential identity resource, then the error would be this (taken from my experiments with external-dns):

`external-dns-595b67dbc4-7pqz7 external-dns time="2022-10-26T17:01:16Z" level=error msg="azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/aks-rg/providers/Microsoft.Network/dnsZones?api-version=2018-05-01: StatusCode=400 -- Original Error: adal: Refresh request failed. Status Code = '400'. Response body: {\"error\":\"invalid_request\",\"error_description\":\"AADSTS70021: No matching federated identity record found for presented assertion. Assertion Issuer: 'https://westeurope.oic.prod-aks.azure.com/bb9ebc7f-b105-4532-95ca-abf3431ff466/c4f4982c-33b9-4f49-b973-8f0095c5c551/'. Assertion Subject: 'system:serviceaccount:external-dns:external-dns'. Assertion Audience: 'api://AzureADTokenExchange'. https://docs.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation\\r\\nTrace ID: d9593b57-d1f7-41f3-9ef7-a236250f6500\\r\\nCorrelation ID: 3bb467a9-eb14-4291-a1b8-8502e75ebfac\\r\\nTimestamp: 2022-10-26 17:01:16Z\",\"error_codes\":[70021],\"timestamp\":\"2022-10-26 17:01:16Z\",\"trace_id\":\"d9593b57-d1f7-41f3-9ef7-a236250f6500\",\"correlation_id\":\"3bb467a9-eb14-4291-a1b8-8502e75ebfac\",\"error_uri\":\"https://login.microsoftonline.com/error?code=70021\"} Endpoint https://login.microsoftonline.com/bb9ebc7f-b105-4532-95ca-abf3431ff466/oauth2/token?api-version=1.0"`

That makes me think that it's better not to implicitly rely on the `if spt != nil` check, we should rather have an explicit way to force cert-manager to use Workload Identity. Then error handling and the code itself would become more straight-forward.

Reply:
Thinking about that more, if the ambient flag is the way to go in cert-manager, then we can use something like this:
```go
// Use Workload Identity if present
if os.Getenv("AZURE_FEDERATED_TOKEN_FILE") != "" {
    return getWorkloadIdentity(env)
}
```
instead of:
```go
spt, err := getWorkloadIdentity(env)
if spt != nil {
    return spt, nil
}
```

To me, that seems to be concise enough, and no errors are suppressed.